gwas-tools.1
GWAS-TOOLS(1)                  User Commands                  GWAS-TOOLS(1)

NAME
       gwas-tools - perform GWAS analysis on phenotypes and genotypes

SYNOPSIS
       gwas-tools [--pheno <phenotypes.csv>] [--geno <genotypes.vcf>]
                  [--out <output_directory>] [--maf <min_maf>]
                  [--count <sample_count>] [--help] [--version] [--debug] [--threads <thread_count>] [--process]

DESCRIPTION
       The gwas-tools command-line tool performs GWAS (Genome-Wide Association
       Study) analysis on phenotypes and genotypes. It allows you to analyze
       genetic associations and identify variants associated with traits of
       interest.

OPTIONS
       --pheno, -p <phenotypes.csv>
              Specify the CSV file containing phenotypes. The first column
              must be sample ID, and the second column must be numeric
              phenotype measurements.

       --geno, -g <genotypes.vcf>
              Specify the VCF file containing genotypes.

       --out, -o <output_directory>
              Specify the output directory for storing the results.

       --maf <min_maf>
              (Optional) Specify the minimum minor allele frequency (between 0
              and 1).

       --count <sample_count>
              (Optional) Specify the number of samples to include.

       --help, -h
              Display the help message.

       --version
              Display the version information.

       --debug
              Print debug messages.

       --threads, -t <thread_count>
              (Optional) Specify the number of threads to use.

       --process
              (Optional) Use the process instead of thread, which is faster but requires more memory.

EXAMPLES
       Perform GWAS analysis on phenotypes and genotypes:
              gwas-tools --pheno phenotypes.csv --geno genotypes.vcf
              --out .

AUTHOR
       Yifei Ding

REPORTING BUGS
       Report bugs to: yifeiding@protonmail.com

COPYRIGHT
       Copyright © 2023 Yifei Ding

SEE ALSO
       More information about GWAS can be found at the following link:
       <https://en.wikipedia.org/wiki/Genome-wide_association_study>

NOTES
       This tool requires Python 3.10 or newer.

BUGS
         No known bugs.

VERSION
       gwas-tools version 0.10

gwas-tools.py
from sqlite3 import DatabaseError
import pandas as pd
import numpy as np
import argparse
import scipy.stats as stats
import matplotlib.pyplot as plt
import sys

VERSION = "1.0.0"


class InvalidFileFormatError(Exception):
    pass


class GenoPhenoMismatch(Exception):
    pass


def read_geno(vcf_file: str):
    """
    Read genotype file
    :param vcf_file: genotype file in VCF format
    :return: genotype data with sample ID and numeric genotype values
    """
    column_name = []
    mapping = {"0|0": 1, "1|0": 2, "0|1": 2, "1|1": 3}
    np.linalg.eig()
    # TODO(Yifei Ding): need error handling for open file
    with open(vcf_file, 'r') as file:
        for line in file:
            if line.startswith(
                    '#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT'):
                column_name = line.split('\t')
                column_name[0] = column_name[0][1:]
                break

    if len(column_name) == 0:
        raise InvalidFileFormatError("Invalid VCF file format")

    try:
        vcf: pd.DataFrame = pd.read_csv(vcf_file, comment="#", sep="\t", names=column_name)
    except pd.errors.ParserError as e:
        raise InvalidFileFormatError("Invalid VCF file format", str(e))

    if vcf.shape[0] == 0:
        raise InvalidFileFormatError("Invalid VCF file format")

    # map genotypes to numeric values 1,2,3
    # TODO: need error handling for outside mapping genotype
    for i in range(9, len(column_name)):
        vcf[column_name[i]] = vcf[column_name[i]].map(mapping)
    # TODO: may need to trim data
    return vcf


def read_pheno(pheno_file: str):
    """
    Read phenotype file
    :param pheno_file: phenotype file
    :return: phenotype data with sample ID and float phenotype measurements
    """

    column_name = ["ID", "Phenotype"]

    try:
        pheno: pd.DataFrame = pd.read_csv(pheno_file, sep="\t", names=column_name)
    except pd.errors.ParserError as e:
        raise InvalidFileFormatError("Invalid phenotype file format", str(e))

    if pheno.shape[0] == 0:
        InvalidFileFormatError("Invalid phenotype file format")

    # TODO: need error handling for non-numeric phenotype
    pheno[column_name[1]] = pheno[column_name[1]].astype(float)

    return pheno


def verify_geno_pheno(geno: pd.DataFrame, pheno: pd.DataFrame):
    """
    Verify that the genotype and phenotype files match in number of samples and sample IDs
    :param geno: genotype data
    :param pheno: phenotype data
    :return: True if genotype and phenotype files match
    """
    # column mismatch
    if geno.shape[1] - 9 != pheno.shape[0]:
        raise GenoPhenoMismatch("Number of samples in genotype and phenotype files do not match")
    for i in range(9, geno.shape[1]):
        if geno.columns[i] != pheno.iloc[i - 9, 0]:
            raise GenoPhenoMismatch("Sample IDs in genotype and phenotype files do not match")
    return True

def generate_plot(data: pd.DataFrame):
    """
    Generate a QQ plot of the expected and observed -log10(p-values)
    :param data: p-value data
    """
    if 'pvalues' not in data.columns:
        raise ValueError("Input DataFrame should contain a 'pvalues' column.")

    num_pvalues = len(data)
    expected_neglogp = -np.log10(np.random.uniform(0,1,num_pvalues))
    expected_neglogp= np.sort(expected_neglogp)
    # Calculate observed -log10(p) values
    observed_neglogp = -np.log10(data['pvalues'])
    observed_neglogp= np.sort(observed_neglogp)

    # Create a new DataFrame with observed and expected -log10(p) values- might have to generate expected values (?)
    sorted_data = pd.DataFrame({'Expected -log10(p)': expected_neglogp,
                         'Observed -log10(p)': observed_neglogp})

    # Plot the expected vs observed -log10(p) values
    plt.figure(figsize=(8, 6))
    plt.scatter(sorted_data['Expected -log10(p)'], sorted_data['Observed -log10(p)'],
                color='b', alpha=0.5)
    plt.plot([min(sorted_data['Expected -log10(p)']), max(sorted_data['Expected -log10(p)'])],
             [min(sorted_data['Observed -log10(p)']), max(sorted_data['Observed -log10(p)'])],
             color='r', linestyle='--')
    plt.xlabel('Expected -log10(p)')
    plt.ylabel('Observed -log10(p)')
    plt.title('GWAS QQ Plot')
    plt.show()

def generate_stats():
    return True


def write_stats():
    return True


def calc_stats(phenotypes: np.vstack, genotypes: np.vstack):
    ## TODO: Use dataframe instead of vstack
    """
    Calculate statistics for GWAS

    :param phenotypes: phenotype data
    :param genotypes: genotype data
    :return: statistics of linear regression
    """
    slopes, intercepts, r_values, p_values, std_errs = stats.linregress(genotypes, phenotypes)
    return {"slopes": slopes, "intercepts": intercepts, "rvalues": r_values, "pvalues": p_values,
            "stderrs": std_errs}

def filter_maf(geno: pd.DataFrame, maf: float):
    """
    Filter genotypes by minor allele frequency
    :param geno: genotype data
    :param maf: minor allele frequency between 0 and 1
    :return: genotype data with minor allele frequency greater than maf
    """
    return geno

def filter_count(geno: pd.DataFrame, count: int):
    """
    Filter genotypes by sample count
    :param geno: genotype data
    :param count: count of samples
    :return: genotype data with sample count greater than count
    """

    return geno


def run_gwas(phenotypes: str, genotypes: str, out: str, maf=None, count=None):
    """
    Run GWAS analysis on phenotypes and genotypes
    :param phenotypes: phenotype file
    :param genotypes: genotype file
    :param out: output directory
    :param maf: minor allele frequency between 0 and 1
    :param count: minimum sample count
    :return: TODO
    """

    print("run gwas")


def main():
    parser = argparse.ArgumentParser(prog='gwas-tools',
                                     description='Perform GWAS analysis on phenotypes and genotypes.')
    parser.add_argument('--pheno', '-p', type=str, help='CSV file containing phenotypes. The first column must be \
    sample ID and the second column must be numeric phenotype measurements.', required=True)
    parser.add_argument('--geno', '-g', type=str, help='VCF file containing genotypes',
                        required=True)
    parser.add_argument('--out', '-o', type=str, help='Output directory for results', required=True)
    parser.add_argument('--maf', type=float,
                        help='Minimum minor allele frequency (between 0 and 1)')
    parser.add_argument('--count', type=int, help='Number of samples to include')
    parser.add_argument('--version', action='version', version='gwas-tools ' + VERSION)

    args = parser.parse_args()
    if args.pheno == '-h' or args.geno == '-h' or args.out == '-h':
        parser.print_help()
    else:
        run_gwas(args.pheno, args.geno, args.out, args.maf, args.count)


if __name__ == '__main__':
    main()
